### PR TITLE
Decompile cut field effect

### PR DESCRIPTION
**Pull request by DizzyEggg**
This was such a fun file to document, especially given how it changed from its Ruby counterpart. I added a couple of nice defines and corrected two functions and would appreciate if the pokeruby's file got updated with these.

### CODE REVIEW

File: src/fldeff_strength.c [hunk 1]
```diff
@@ -14,11 +14,11 @@ static void sub_8145E74(void);
 // text
 bool8 SetUpFieldMove_Strength(void)
 {
-    if (npc_before_player_of_type(87) == TRUE)
+    if (SetLastTalkedObjectInFrontOfPlayer(87) == TRUE)
```
Comment (Diegoisawesome) [inline src/fldeff_strength.c]: Use define

File: src/rom6.c [hunk 2]
```diff
@@ -121,14 +121,14 @@ bool8 SetUpFieldMove_RockSmash(void)
...
     {
         gSpecialVar_Result = GetCursorSelectionMonId();
-        gUnknown_03005DB0 = FieldCallback_Teleport;
-        gUnknown_0203CEEC = sub_8179834;
+        gFieldCallback2 = FieldCallback_PrepareFadeInFromMenu;
+        gPostMenuFieldCallback = sub_8179834;
         return TRUE;
     }
-    else if (npc_before_player_of_type(0x56) == TRUE)
+    else if (SetLastTalkedObjectInFrontOfPlayer(0x56) == TRUE)
```
Comment (Diegoisawesome) [inline src/rom6.c]: Use define

File: src/fldeff_cut.c [hunk 3]
```diff
@@ -1,19 +1,684 @@
...
+                while (1)
+                {
+                    if (sHyperCutStruct[i].unk2[j] == 0)
+                        break;
+                    if (cutTiles[(u8)(sHyperCutStruct[i].unk2[j] - 1)] == FALSE)
+                    {
+                        tileCuttable = FALSE;
+                        break;
+                    }
+                    if (++j >= 2)
```
Comment (Diegoisawesome) [inline src/fldeff_cut.c]: Do while?

File: src/fldeff_cut.c [hunk 4]
```diff
@@ -1,19 +1,684 @@
...
+    FieldEffectStart(FLDEFF_CUT_GRASS);
+}
+
+bool8 FldEff_CutGrass(void)
+{
+    s16 x, y;
+    u8 tileBehavior;
+    u8 i;
+
+    for (i = 0, PlaySE(SE_W015), PlayerGetDestCoords(&gPlayerFacingPosition.x, &gPlayerFacingPosition.y); i < CUT_HYPER_AREA; i++)
```
Comment (Diegoisawesome) [inline src/fldeff_cut.c]: I think you should move the two functions out of the for statement (and if necessary, the i = 0 if it comes before those in the code)

Comment (DizzyEggg) [reply src/fldeff_cut.c]: Copied directly from pokeruby.

File: src/fldeff_cut.c (hunk 3 above)
Comment (DizzyEggg) [reply src/fldeff_cut.c]: Sure.

File: src/rom6.c (hunk 2 above)
Comment (DizzyEggg) [reply src/rom6.c]: done

File: src/fldeff_strength.c (hunk 1 above)
Comment (DizzyEggg) [reply src/fldeff_strength.c]: done

File: src/fldeff_cut.c (hunk 4 above)
Comment (Diegoisawesome) [reply src/fldeff_cut.c]: Still, if it can be done better than pokeruby...

File: src/rom6.c (hunk 2 above)
Comment (Diegoisawesome) [reply src/rom6.c]: Rename `SetLastTalkedObjectInFrontOfPlayer` to `CheckObjectGraphicsInFrontOfPlayer`

Comment (Diegoisawesome) [reply src/rom6.c]: Egg pls

**Pull request merged**